modules/videoio/src/cap_openni2.cpp

```diff
@@ -191,72 +193,113 @@ openni::VideoMode CvCapture_OpenNI2::defaultStreamOutputMode(int stream)
+    openni::Status status;
     status = openni::OpenNI::initialize();
-
     if (status != openni::STATUS_OK)
     {
         CV_Error(CV_StsError, cv::format("Failed to initialize:", openni::OpenNI::getExtendedError()));
         return;
     }
 
-    // find appropriate device URI
-    openni::Array<openni::DeviceInfo> ldevs;
-    if (index > 0)
+    const char* deviceURI = openni::ANY_DEVICE;
+    bool needColor = true;
+    bool needIR = true;
+    if (index >= 0)
+    {
+        int deviceType = DEVICE_DEFAULT;
+        if (index >= 10)
+        {
+            deviceType = index / 10;
+            index %= 10;
+        }
+        // Asus XTION and Occipital Structure Sensor do not have an image generator
+        needColor = (deviceType != DEVICE_ASUS_XTION);
+
+        // find appropriate device URI
+        openni::Array<openni::DeviceInfo> ldevs;
+        if (index > 0)
+        {
+            openni::OpenNI::enumerateDevices(&ldevs);
+            if (index < ldevs.getSize())
+                deviceURI = ldevs[index].getUri();
+            else
+            {
+                CV_Error(CV_StsError, "OpenCVKinect2: Device index exceeds the number of available OpenNI devices\n");
+                openni::OpenNI::shutdown();
+                return;
+            }
+        }
+    }
+    else
     {
-        openni::OpenNI::enumerateDevices(&ldevs);
-        deviceURI = ldevs[index].getUri();
+        deviceURI = filename;
     }
 
     status = device.open(deviceURI);
-    if( status != openni::STATUS_OK )
+    if (status != openni::STATUS_OK)
     {
-        CV_Error(CV_StsError, cv::format("OpenCVKinect: Device open failed see: %s\n", openni::OpenNI::getExtendedError()));
+        CV_Error(CV_StsError, cv::format("OpenCVKinect2: Failed to open device: %s\n", openni::OpenNI::getExtendedError()));
         openni::OpenNI::shutdown();
         return;
     }
 
     status = toggleStream(CV_DEPTH_STREAM, true);
-    // Asus XTION and Occipital Structure Sensor do not have an image generator
-    if (deviceType != DEVICE_ASUS_XTION)
+    if (needColor)
         status = openni::Status(status | toggleStream(CV_COLOR_STREAM, true));
+    if (needIR)
+        status = openni::Status(status | toggleStream(CV_IR_STREAM, true));
     if (status != openni::STATUS_OK)
     {
         openni::OpenNI::shutdown();
         return;
     }
 
-    if (!readCamerasParams())
-    {
-        CV_Error(CV_StsError, cv::format("CvCapture_OpenNI2::CvCapture_OpenNI2 : Could not read cameras parameters\n"));
-        return;
-    }
+    readCamerasParams();
 
+    setProperty(CV_CAP_PROP_OPENNI_REGISTRATION, 1.0);
 
-    outputMaps.resize( outputMapsTypesCount );
+    // default for Kinect2 camera
+    setProperty(CV_CAP_PROP_OPENNI2_MIRROR, 0.0);
 
     isContextOpened = true;
+}
 
-    setProperty(CV_CAP_PROP_OPENNI_REGISTRATION, 1.0);
+CvCapture_OpenNI2::~CvCapture_OpenNI2()
+{
+    for (int i = 0; i < CV_MAX_NUM_STREAMS; ++i)
```

Comment:
> i < CV_MAX_NUM_STREAMS

It is `std::vector<>` now. So it is better to use .size() method instead of some "OpenCV" constant.